arbiter/drivers/onedrive.cpp

```diff
@@ -0,0 +1,292 @@
+            return std::vector<std::string>();
+        }
+
+        //parse list
+        const json list(obj.at("value"));
+        for (auto& i: list.items())
+        {
+            const auto data(i.value());
+            const std::string fileName(data.at("name").get<std::string>());
+            const std::string filePath(path + "/" + fileName);
+
+            result.push_back(filePath);
+            if (data.contains("folder") && recursive)
+            {
+                //restart process with new file head
+                const auto children(processList(filePath, recursive));
+
+                //add result of children processes to the parent
+                result.insert(result.end(), children.begin(), children.end());
+            }
+        }
+
+        //check for link to next set
+        if (obj.contains("@odata.nextLink"))
+            pageUrl = obj.at("@odata.nextLink");
+        else
+            break;
+
+    } while (true);
+
+    return result;
+}
+
+std::vector<std::string> OneDrive::glob(std::string path, bool verbose) const
+{
+    path.pop_back();
+    bool recursive(path.back() == '*');
+    if (recursive)
+        path.pop_back();
+
+    if (path.back() == '/')
+        path.pop_back();
+
+    return processList(path, recursive);
+}
+
+bool OneDrive::get(
+        const std::string path,
+        std::vector<char>& data,
+        const http::Headers userHeaders,
+        const http::Query query) const
+{
+    Resource resource(path);
+    http::Headers headers(m_auth->headers());
+    headers["Content-Type"] = "application/octet-stream";
+    headers.insert(userHeaders.begin(), userHeaders.end());
+
+    drivers::Https https(m_pool);
+    const auto res(https.internalGet(resource.binaryEndpoint(), headers));
+
+    if (!res.ok()) {
+        std::cout <<
+            "Failed get - " << res.code() << ": " << res.str() << std::endl;
+        return false;
+    }
+
+    data = res.data();
+    return true;
+}
+
+OneDrive::Auth::Auth(std::string s) {
+    const auto config = json::parse(s);
+    m_token = config.at("access_token").get<std::string>();
+    m_refresh = config.at("refresh_token").get<std::string>();
+    m_redirect = config.at("redirect_uri").get<std::string>();
+    m_id = config.at("client_id").get<std::string>();
+    m_secret = config.at("client_secret").get<std::string>();
+    m_expiration = Time().asUnix();
+}
+
+std::unique_ptr<OneDrive::Auth> OneDrive::Auth::create(const std::string s)
+{
+    return makeUnique<Auth>(s);
+}
+
+void OneDrive::Auth::refresh()
+{
+    //only refresh if we get within 2 minutes of the token ending
+    const auto now(Time().asUnix());
+    if (m_expiration - now > 120)
+        return;
+
+    http::Pool pool;
+    drivers::Https https(pool);
+    http::Headers headers({ });
+
+    headers["Accept"] = "application/json";
+    headers["Content-Type"] = "application/x-www-form-urlencoded";
+
+    json bodyJson = {
```

Comment:
No need to go through JSON here.  I'd use the `buildQueryString` utility that already exists in this project, although looking at it you'd need to modify it to not put the leading `?` before its result (and update any callers to match the change in behavior).